Use WinAPI concurrency primitives on Windows ports (remove winpthreads)

MisterDA opened the pull request:
The goal is to remove winpthreads (a library emulating pthreads on top of _old_ Windows APIs), and use _modern_ Windows APIs in both the MSVC and MinGW-w64 ports.

~It would be great if this PR could be considered for inclusion in 5.3: although we were already using the system winpthreads for the MinGW-w64 port, we introduced it as a submodule during the 5.3 development cycle to support the MSVC port. If we could remove it now, end users won't ever (unknowingly) depend on it.~ This PR is also based on #13352.

One commit removes winpthreads (breaking the OCaml build on the MSVC/clang-cl ports) and another implements the `caml_plat_*` and `st_*` functions required to abstract over pthreads and Windows concurrency APIs. 

Note that Windows [Mutex Objects](https://learn.microsoft.com/en-us/windows/win32/sync/mutex-objects) cannot be used with condition variables. The right pairing consists of [Slim Reader/Writer Locks](https://learn.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks) (SRW locks) and [condition variables](https://learn.microsoft.com/en-us/windows/win32/sync/condition-variables). SRW locks support shared locking (only readers) and exclusive locking (one writer). I've chosen to use exclusive locking in all cases. Windows condition variables have the problem that they cannot be statically initialized.

Both pthreads and WinAPI are quite similar, except for a few points that I've noted:
1. The type of a function that starts a thread. They return a `void *` pointer on pthreads, and an `unsigned int` on Windows, but it seems not to be a concern for us.
2. We use pthread `ERRORCHECK` mutexes. [POSIX states](https://pubs.opengroup.org/onlinepubs/9799919799/functions/pthread_mutex_lock.html):

   > The `pthread_mutex_lock()` function shall fail if:
   > [`EDEADLK`]
   >   The mutex type is `PTHREAD_MUTEX_ERRORCHECK` and the current thread already owns the mutex.

   The [Windows documentation states](https://learn.microsoft.com/en-us/windows/win32/sync/slim-reader-writer--srw--locks):

   > Exclusive mode SRW locks cannot be acquired recursively. If a thread tries to acquire a lock that it already holds, that attempt will […] deadlock (for `AcquireSRWLockExclusive`).

   Deadlocks caused by recursive attempts to lock cannot be detected on Windows. We cannot either detect double unlocks, or attempts to unlock a lock owned by another thread. This would revert the `Stdlib.Mutex` module guarantees to those of before OCaml 4.12 _on Windows only_.
   
   I've thus chosen to implement so-called `ERRORCHECK` mutexes with a pair of a SRWLock, and the thread id of the owner, accessed atomically. The value 0 is used when the thread is not owned, and is a valid sentinel value.

3. A Windows thread is "detached" if all handles referring to the thread have been closed: it cannot be joined (waited on in Windows terminology) anymore but is not ended. pthreads has a thread attribute to control the detached state.

I'd like to thank my colleagues at Tarides who have helped my write (and debug!) this code.

dra27 committed:
Don't clobber GetLastError in Win32 systhreads

See also PR 10220 and issue 8857.
SetLastError is called when accessing the TLS value This_thread.

(cherry picked from commit 93843575809433ea50e0406d56c0681287411f1b)

nojb commented:
Thanks for the PR! Have you made any benchmarks to have an idea of what is the performance impact of this change?

> It would be great if this PR could be considered for inclusion in 5.3

I'm afraid that ship has sailed :) 5.3 is in bug-fixing mode, and this PR is much too big to slip it in (cc @Octachron).

gasche commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
 }
 
-#include "st_pthreads.h"
+/* POSIX thread implementation of the "st" interface */
```
Is this comment still accurate?

dustanddreams replied:
In this first commit, things are simply moved around, it wouldn't make sense to also change the contents at this point.

MisterDA replied:
The first commit is supposed to help reviewers by simplifying the diff from pthreads. I can remove this comment in the latter commits.

gasche commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+  caml_release_domain_lock();
+
+  do {
+    /* Note: the POSIX spec prevents the above signal from pairing with this
```
The comment refers to the POSIX spec, but this is using a Windows API function, does the POSIX spec actually apply?

dustanddreams replied:
It does, because the OCaml runtime is written to rely upon POSIX semantics, which the windows implementation has to implement/mimic.

gasche reviewed (commented):
Some drive-by comments from just skimming the PR.

I think that it would make sense to defer the ERRORCHECK emulation to a later PR, to make the "core" of the PR smaller and easier to review. (The ERRORCHECK logic is well separated, but non-trivial, it will consume some review energy.)

What has been your testing strategy so far? I suppose that the testsuite passes, but maybe there are other interesting tests that would exercise the thread machinery?

dra27 commented:
Full review to follow, but I disagree about splitting the PR - the errorcheck part is in a separate commit, so not much easier to review in another PR and somewhat more importantly, the two PRs would still have to be merged at the same time

gasche commented:
Is it a big issue if the Windows port does not implement ERRORCHECK and deadlocks instead? My impression was that ERRORCHECK is there to catch programming errors (in particular, this is not something that users should ever rely on dynamically in the program behavior), and it seems reasonable to ask people to use another setup if they want certain kind of debugging (this is already what we do for the various sanitizers, that I believe are also unavailable on MSVC). My intuition is that it would be okay to have our MSVC port revert to pre-ERRORCHECK behavior for a release.

ghost commented:
The debug runtime relies upon the ability to detect locking recursion.

MisterDA commented:
Prior discussion on the pros and cons of errorcheck mutexes for OCaml 4 can be found at #9757 and #9846. I even considered merging the errorcheck patch into _Implement concurrency primitives using WinAPI_ because I thought the diff with pthreads might be easier to follow.

> Have you made any benchmarks to have an idea of what is the performance impact of this change?

No, I haven't done any benchmarks. I think we lack infrastructure to benchmark the compiler on Windows, but I'd love to be proved wrong!

>> It would be great if this PR could be considered for inclusion in 5.3
>
> I'm afraid that ship has sailed :) 5.3 is in bug-fixing mode, and this PR is much too big to slip it in

Dura lex, sed lex.

> What has been your testing strategy so far? I suppose that the testsuite passes, but maybe there are other interesting tests that would exercise the thread machinery?

Building the compiler and running the testsuite caught the early bugs. I used the test [`testsuite/tests/lib-threads/mutex_errors.ml`](https://github.com/ocaml/ocaml/blob/trunk/testsuite/tests/lib-threads/mutex_errors.ml) to validate my errorcheck code. We ran multicoretests for [MinGW-w64](https://github.com/ocaml-multicore/multicoretests/actions/runs/10671106341/job/29576298656) and [MSVC](https://github.com/ocaml-multicore/multicoretests/actions/runs/10671106338/job/29576298767) (it's [going to be easier to test a specific branch](https://github.com/ocaml-multicore/multicoretests/pull/473) soon) and found that two negative tests (supposed to fail) were succeeding.

Quoting @jmid on the results:

> The `Systhread` mode of multicoretests does not work as well as the `Domain` mode unfortunately.
To ensure that both work reasonably, we have added 'negative tests' that hammer an unprotected `int ref` or `int64 ref` from either (a) two concurrent threads, or (b) two parallel domains, and expect to discover some misbehaviour across sufficiently many repetitions.
>
> For an `int ref` there's little allocation to trigger a context switch, so we consider those tests positive (don't expect to observe misbehaviour).
> For `int64 ref` there are more boxed allocations, and we are usually able to trigger misbehaviour with a native code compiler.
> With your PR to eliminate pthreads we are no longer able to do so on MSVC and MinGW:
>
>     [✗] 5000    0    0 5000 / 5000   160.4s STM int64 ref test with Thread
>    
>     [...]
>
>     [✗] 15000     0     0 15000 / 15000   398.7s Lin.Internal ref int64 test with Thread
> It raises an eyebrow, but it is not a problem per se, just a bit strange!

dustanddreams commented on otherlibs/systhreads/st_stubs.c:
```diff
@@ -699,7 +701,7 @@ static void * caml_thread_tick(void * arg)
     caml_interrupt_self();
   }
-  return NULL;
+  return 0;
```
I think it would be nice to make all thread routines return an explicitly cast value, i.e.
```suggestion
  return (CAML_THREAD_FUNCTION)0;
```

But maybe that's too pedantic.

MisterDA replied:
It's a bit pendantic, but doesn't hurt. I'll add this where needed.

MisterDA replied:
Ah, of course this doesn't work as-is because of the `WINAPI` specifier ;)

dustanddreams replied:
Damn. No casts, then.

MisterDA replied:
Note how it is defined:
```c
#ifdef _WIN32
int caml_plat_thread_create(caml_plat_thread *restrict,
                            const caml_plat_thread_attr *restrict,
                            unsigned (WINAPI *)(void *),
                            void *restrict);
#define CAML_THREAD_FUNCTION unsigned WINAPI
#else
/* ... */
#define CAML_THREAD_FUNCTION void *
#endif
```

The `WINAPI` macro needs to be put inside the parentheses for the function pointer type!

Applied to `return (CAML_THREAD_FUNCTION) 0;`, MSVC complains:

```
runtime/domain.c(1095): warning C4229: anachronism used: modifiers on data are ignored
runtime/domain.c(1278): warning C4229: anachronism used: modifiers on data are ignored
```

I think it'd be too heavy introducing a new macro without the `WINAPI` part, and also ugly to do (if I were to remove the macro):

```c
static
#ifdef _WIN32
unsigned WINAPI
#else
void *
#endif
my_thread_func(void *arg) {
  /* ... */
#ifdef _WIN32
  return 0;
#else
  return NULL;
#endif
}
```

With your permission, I won't include this change; unless you have a better idea on how to write it.

As a reminder, `WINAPI` is a macro that evaluates to [`__stdcall`](https://learn.microsoft.com/en-us/cpp/cpp/stdcall?view=msvc-170), a Microsoft-specific modifier specifying the calling convention used to call Win32 API functions. It only has a meaning on x86 and is ignored on other architectures.

dustanddreams replied:
No, that's fine to rely upon the implicit cast from `int` to `void *`.

dustanddreams commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+    );
+  if (thr == 0) {
+    return errno;
+  } else if (res == NULL) {
```
I think chaining the tests like this is ugly. I would prefer the test for failure to be separated from the test to decide whether to detach the thread or not, as these are completely unrelated operations. In other words:
```suggestion
  }
  
  if (res == NULL) {
```

MisterDA replied:
Makes sense, thanks!

dustanddreams commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+static void st_masterlock_acquire(st_masterlock *m)
+{
+  AcquireSRWLockExclusive(&m->lock);
+  while (m->busy) {
```
I am worried an overzealous compiler (i.e. `clang-cl`) could optimize this into `if (m->busy) while (1)`, because `busy` is marked neither `volatile` nor `atomic`.

I know this logic was already present in the existing code, but I'm worried nevertheless.

MisterDA replied:
I've marked both boolean fields as atomic, and also in the pthreads code.

eutro replied:
Surely the compiler can't optimise this away, because the loop accesses atomics and a condition variable. Am I missing something?

MisterDA replied:
Indeed. I submitted the change to mark these fields as atomic in https://github.com/ocaml/ocaml/pull/13558, and during the discussion there the same conclusion was reached. In the meantime, I had applied the atomic qualifier here then reverted the change. This PR _does not_ make the fields atomic.

dustanddreams commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+  m->busy = false;
+  atomic_fetch_add(&m->waiters, +1);
+  WakeConditionVariable(&m->is_free);
+  /* releasing the domain lock but not triggering bt messaging
+     messaging the bt should not be required because yield assumes
```
```suggestion
  /* releasing the domain lock but not triggering bt messaging.
     Messaging the bt should not be required because yield assumes
```

dustanddreams commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+  /* releasing the domain lock but not triggering bt messaging
+     messaging the bt should not be required because yield assumes
+     that a thread will resume execution (be it the yielding thread
+     or a waiting thread */
```
```suggestion
     or a waiting thread) */
```

dustanddreams commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,242 @@ Caml_inline void st_msleep(int msec)
+  WakeConditionVariable(&m->is_free);
+  /* releasing the domain lock but not triggering bt messaging
+     messaging the bt should not be required because yield assumes
+     that a thread will resume execution (be it the yielding thread
```
```suggestion
     that a thread will resume execution (either the yielding thread
```

gasche commented:
> We ran multicoretests for [MinGW-w64](https://github.com/ocaml-multicore/multicoretests/actions/runs/10671106341/job/29576298656) and [MSVC](https://github.com/ocaml-multicore/multicoretests/actions/runs/10671106338/job/29576298767) (it's https://github.com/ocaml-multicore/multicoretests/pull/473 soon)

This is what I was hoping to hear, and I'm convinced that this is a strong argument in favor of correctness of the implementation. Thanks!

dustanddreams commented on runtime/sync_win32.h:
```diff
@@ -0,0 +1,133 @@
+/*                                                                        */
+/*                                 OCaml                                  */
+/*                                                                        */
+/*                         Antonin Decimo, Tarides                        */
```
```suggestion
/*                         Antonin Décimo, Tarides                        */
```

MisterDA replied:
I'd like to, but I think we're only allowed ASCII. IIRC there's a Hygiene check that prevents this. TBH I'm not even sure my name officially has an accent.

dustanddreams replied:
Ah, right.

shym replied:
check-typo won’t like that, will it?

dra27 replied:
Just FTR, this was fixed years ago (by me!) in #1287. cf. https://github.com/ocaml/ocaml/blob/e39fcc48b1b573ee49c86843cfd02737977fe682/utils/compression.mli#L5

shym reviewed (approved):
I’ve reviewed the range-diff with the previous iteration of this PR and this looks good to me.

MisterDA commented on runtime/sync_win32.h:
```diff
@@ -0,0 +1,133 @@
+    m->owner_tid = 0;
+    if (SleepConditionVariableSRW(c, &m->lock, INFINITE,
+                                  0 /* exclusive */)) {
+      m->owner_tid = GetCurrentThreadId();
```
```suggestion
      m->owner_tid = self_tid;
```

MisterDA commented on runtime/win32.c:
```diff
@@ -1154,6 +1156,151 @@ void caml_plat_mem_unmap(void* mem, uintnat size)
+    mut->owner_tid = 0;
+    if (SleepConditionVariableSRW(cond, &mut->lock, INFINITE,
+                                  0 /* exclusive */)) {
+      mut->owner_tid = GetCurrentThreadId();
```
```suggestion
      mut->owner_tid = self_tid;
```

MisterDA commented on Changes:
```diff
@@ -7,6 +7,14 @@ Working version
+
+- #13416: Implement concurrency primitives using WinAPI instead of
+  winpthreads on Windows.
+  (Antonin Décimo, review by Samuel Hym and Miod Vallat)
```
```suggestion
  (Antonin Décimo, review by Samuel Hym, Miod Vallat, and Gabriel Scherer)
```

MisterDA commented on otherlibs/systhreads/st_win32.h:
```diff
@@ -24,7 +24,241 @@ Caml_inline void st_msleep(int msec)
+
+typedef struct st_event_struct {
+  SRWLOCK lock;                 /* to protect contents */
+  bool status;                  /* false = not triggered, true = triggered */
```
@dustanddreams should this field also be marked atomic or volatile? There's a similar loop pattern in `st_event_wait`.

dustanddreams replied:
Yes, I think so.

dra27 commented:
> Is it a big issue if the Windows port does not implement ERRORCHECK and deadlocks instead? My impression was that ERRORCHECK is there to catch programming errors (in particular, this is not something that users should ever rely on dynamically in the program behavior), and it seems reasonable to ask people to use another setup if they want certain kind of debugging (this is already what we do for the various sanitizers, that I believe are also unavailable on MSVC). My intuition is that it would be okay to have our MSVC port revert to pre-ERRORCHECK behavior for a release.

It's not just the MSVC port - it's _both_ Windows ports, which would regress the behaviour of the mingw-w64 port. It's not necessarily debugging, it's correctness - it'd be a little bit like not having array bounds checking 😊 It was deemed important enough in #9757 / #9846 to go to some effort to make the Windows behaviour the same and it wasn't regressed in 5.0, so I (strongly) think it should continue to matter.

gasche commented:
Okay, okay! As long as people review this carefully, either approaches are fine.

MisterDA commented:
> > > It would be great if this PR could be considered for inclusion in 5.3
> >
> > I'm afraid that ship has sailed :) 5.3 is in bug-fixing mode, and this PR is much too big to slip it in
> 
> Dura lex, sed lex.

There's been a handful of patches to upstream winpthreads since we introduced it during the 5.3 development cycle. If this PR isn't picked for this release, we should at least sync with mingw-w64 upstream before the release is cut. (and perhaps FlexDLL too, while we're at it).

Octachron commented:
> If this PR isn't picked for this release

This PR will not be picked for the 5.3 release, it is far too late for being even considered for inclusion.

xavierleroy commented:
I need to be convinced this is not a regression compared to just using winpthreads, or a more modern implementation of the POSIX API for Win32.  There are benefits to coding to a single API.

MisterDA commented:
> I need to be convinced this is not a regression compared to just using winpthreads, or a more modern implementation of the POSIX API for Win32. There are benefits to coding to a single API.

The Windows concurrency API is obviously *not* pthreads, but they map rather cleanly. We use threads, mutexes, condition variables, and thread-specific data, and for each of these Windows has had a (relatively straightforward) equivalent since Vista. Implementing this patch series, I felt I mostly needed to care about correctly mapping error code and return values, and implementing the error check feature of mutexes.

winpthreads has to keep working with [Windows 98](https://github.com/mingw-w64/mingw-w64/commit/cb7f42e05b94cbb1d94edcebdbfb47b652a45484). It thus has to use the old [Interlocked Variable Access](https://learn.microsoft.com/en-us/windows/win32/sync/interlocked-variable-access) from the WinAPI or [Legacy __sync Built-in Functions for Atomic Memory Access](https://gcc.gnu.org/onlinedocs/gcc/_005f_005fsync-Builtins.html). Upstream is active and healthy, but there's no strong guarantee that they'll proactively test changes for MSVC compatibility.

There's also the [pthread-win32](https://github.com/GerHobbelt/pthread-win32)/pthreads4w project. Like winpthreads, it also uses the Interlocked Variable Access, but advertises support for MSVC. It seems to support more from POSIX 2008 than winpthreads.

IMHO both projects suffer from having to re-implement the concurrency primitives using old APIs. They probably cannot switch to C11 atomics either. This considerably raises the maintenance burden and the barrier (ha!) of entry. By switching to the WinAPI we have a much lighter and inspectable layer of code. We also already have a layer of abstraction above the pthreads API, which I extended to cover all the features we require from it.

See #12954 for a justification by @dra27 of why we used winpthreads, and also why it was deemed temporary.

> The first port of the Multicore OCaml project to Windows in https://github.com/ocaml-multicore/ocaml-multicore/pull/173 for mingw-w64 used this library as a "temporary" workaround, as it allowed all the existing pthreads code (right up to linking with `-lpthread`) to be reused, and allowed efforts to be focused instead on the memory management code, timers and other things which had to be ported. The use of it in 2018 was regarded as temporary for two reasons, principally that we knew that the library was mingw-w64 only, so MSVC support would be difficult.
>
> It is now worth mentioning the _second reason_ for regarding the use of winpthreads as temporary. It has always been @stedolan's intention that OCaml's use of pthreads should be limited - i.e. the runtime eschews "exotic" features of pthreads, mainly for portability concerns. We strongly believe that we should be able to reimplement OCaml's use of pthreads in a more abstract way (exactly as the systhreads library does in 4.x), and therefore implement both the mingw-w64 and MSVC ports without the use of winpthreads. Further, we'd expect to be able to be more performant, simply because our own primitives would not need to be worrying about the full weight of the POSIX spec for pthreads, which of course winpthreads has to.

Hope this helps.

MisterDA commented:
I've looked at some other languages runtime and standard libraries.

I found some C code for Golang that's clearly using the WinAPI, and the Linux equivalent is using pthreads.
https://github.com/golang/go/blob/master/src/runtime/cgo/gcc_libinit_windows.c
Rust stdlib mutexes are implemented with SRWLocks.
https://github.com/rust-lang/rust/blob/master/library/std/src/sys/sync/mutex/windows7.rs
There's also mention that the MSVC C++ STL uses them too to implement `std::shared_mutex`.
Python's condition variables are also implemented natively.
https://github.com/python/cpython/blob/main/Python/condvar.h

MisterDA commented:
Rebased after the end of the cleanup-at-exit effort with #12964.

gasche commented:
Consensus from the maintainer meeting: okay to move forward, @dra27 volunteers to review it.

eutro commented on runtime/caml/platform.h:
```diff
@@ -452,6 +502,45 @@ CAMLextern CAMLthread_local int caml_lockdepth;
 
+#ifdef _WIN32
+
+Caml_inline void caml_plat_lock_blocking(caml_plat_mutex* m)
+{
+  DWORD self_tid = GetCurrentThreadId();
+  if (m->owner_tid != self_tid) {
+    AcquireSRWLockExclusive(&m->lock);
+    m->owner_tid = self_tid;
+    DEBUG_LOCK(m);
+  } else {
+    check_err("lock", EDEADLK);
+  }
+}
```
It seems unfortunate that `ERRORCHECK`-related code should be essentially duplicated in three different places (here, `sync_mutex_lock`, and `caml_plat_lock_non_blocking_actual`, and others). Surely they can be extracted out to functions that directly mirror the pthreads behaviour? e.g.

```c
Caml_inline int caml_plat_lock_errorcheck(caml_plat_mutex* m)
{
  DWORD self_tid = GetCurrentThreadId();
  if (m->owner_tid != self_tid) {
    AcquireSRWLockExclusive(&m->lock);
    m->owner_tid = self_tid;
    return 0;
  } else {
    return EDEADLK;
  }
}
```

The code isn't entirely trivial: it took me a second to convince myself that the `m->owner_tid` check is _not_ a time-of-check to time-of-use error, for instance.

MisterDA replied:
I agree with you, but considering that the duplication between sync and plat modules existed before this PR, and although I'm tempted to try to merge the code, I'm reluctant to do it here or at all (and I would appreciate a greenlight from a core dev before diving in).

nojb replied:
As the duplication existed before this PR, let us leave this potential improvement for a separate PR as well, to keep this one as focused as possible.

MisterDA commented:
Thanks for looking at this @eutro!
In the meantime, I've rebased this PR and split out the part introducing `caml_plat_thread` and `caml_plat_thread_attr` to its own commit, to help reviewing.

nojb commented:
What is the status of this PR? It seems to be dependency of a number of other PRs #13348 (and transitively, #13360 #13355 etc).

MisterDA commented:
It is awaiting more reviews from wizards. 
@dustanddreams and @shym reviewed an early version, but I don't think it counts anymore as a an approval. @eutro reviewed this code somewhat recently, and if I'm not mistaken, approved the PR and suggested a refactor. Unfortunately, none of these hackers are core maintainers. @dra27 committed to a review after he lands from the relocatable ocaml projets. We're never short of eyes, so, @nojb if you've heard of a Windows or pthread expert around here these days, maybe they could have a look at it?

My opinion is that the Windows api maps rather nicely to pthread. The error handling differs, I've added a custom but I hope standard implementation of error check mutexes, and there's just a dash of really low-level Windows.

nojb commented:
> We're never short of eyes, so, @nojb if you've heard of a Windows or pthread expert around here these days, maybe they could have a look at it?

I am neither, but I will try to take a look anyway.

gasche commented:
I haven't seen an explicit approval by @eutro (the meaning being something like "I've looked at this in details, I don't see any remaining issue, and I think this should be merged"), I would be happy to approve on their behalf if that was the case.

gasche commented:
I'd like another discussion (online or offline) with @xavierleroy about whether he agrees with the design.

gasche commented:
I forgot to discuss this with @xavierleroy, and I'm afraid this has stalled the PR.

My current thinking is as follows:

- The current use of winpthreads makes code somewhat more portable, but it also comes at a cost in terms of maintenance.
- The PR proposes a newer approach that could give a cleaner / more specialized Windows implementation, but this is all new code that probably has a few bugs left that will be found over time.
- The cost of the maintenance of the winpthreads approach and the cost of the maintenance of the new approach will be paid by the same people in practice, @dra27 and @MisterDA, maybe @nojb.
- If the people who are going to maintain the Windows support are in favor of the change, I would support trusting their preference and merging the PR. But one could reasonably worry that the time of @dra27 and @MisterDA could be scarce to look at these in the future, in which case the maintenance burden would probably fall to @nojb.

@nojb, you have not expressed an opinion on this PR. Do you think that this approach is reasonable?

nojb commented:
> @nojb, you have not expressed an opinion on this PR. Do you think that this approach is reasonable?

The approach is certainly reasonable.

This PR is replacing the Windows-only compatibility layer `winpthreads` that is not written by us by a compatibility layer for both POSIX and Windows, but written by us, which arguably simplifies matters (as we can design something specifically for our needs).

All in all, I am in favour of going ahead (assuming @MisterDA and @dra27 are still in favour as well :))

nojb reviewed (approved):
I have only looked cursorily at the code, but I agree with the approach and I am approving on the strength of existing reviews.

LGTM

gasche commented:
I propose to go ahead and merge then. (Xavier mentioned that he needed convincing, but I see that later on we discussed this in a dev-meeting and decided that it was acceptable.)

@MisterDA, when you have time, could you rebase to fix the Changes conflict and rerun the CI tests? (I doubt much has changed that touches this part, but we never know.)

MisterDA commented:
Earlier, Xavier wrote:

> I need to be convinced this is not a regression compared to just using winpthreads, or a more modern implementation of the POSIX API for Win32. There are benefits to coding to a single API.

For the sake of exhaustiveness, I know 3 implementations of pthreads for Windows. One is [winpthreads](https://github.com/mingw-w64/mingw-w64/tree/master/mingw-w64-libraries/winpthreads), maintained by the mingw-w64 project. It was forked from pthread-win32 (aka pthreads4w), whose most active fork seems to be https://github.com/GerHobbelt/pthread-win32. Both are now compatible with MSVC and mingw-w64. Finally, there's [MCF Gthread](https://github.com/lhmouse/mcfgthread) from a mingw-w64 maintainer, which provides a C&NoBreak;+&NoBreak;+11 threads implementation with a patched GCC/mingw-w64. It claims to be more efficient than winpthreads and the WinAPI.

pthread-win32 seems to be a 1-person project. MCF Gthread isn't adaptable to our needs (doesn't support MSVC, not clear if it provides pthreads or just the C&NoBreak;+&NoBreak;+ library). winpthreads has the drawback that it has to support really old version of Windows (amongst other ewws it's written in C89 and uses the Interlocked Variable Access API), and some of its code is quite arcane. For instance, the library needs to [setup thread-local storage callbacks](https://github.com/mingw-w64/mingw-w64/blob/2e87f03fc17adf6a1a4f94d6793a4f5ae74fdc89/mingw-w64-libraries/winpthreads/src/thread.c#L523-L549) to initialize itself alongside the CRT. This is what Microsoft says [about it](https://learn.microsoft.com/en-us/cpp/c-runtime-library/crt-initialization?view=msvc-170#linker-features-for-initialization):
>  We don't recommend this Microsoft-specific technique, and it may break in a future release. We've documented it only to keep you from creating code that's broken in hard-to-diagnose ways.

and it's brittle when one tries fancy link-time optimization ([my bug report](https://developercommunity.visualstudio.com/t/Whole-program-optimization-and-optimized/10531811)).

From the commit log of pthread-win32, it seems that the author has partly tried to reimplement pthreads on top of the WinAPI, which is essentially what my PR is doing (for the subset of pthreads that we care about).

We could have switched to the C11 concurrency library but it is optional, and not all platforms provide it. It also has subtle differences with the pthreads API. Some platforms provide pthreads, but not the C11 library. Some platforms provide the C&NoBreak;+&NoBreak;+11 concurrency library, but not the C11 concurrency library. For instance, Microsoft now provides the C11 concurrency library with MSVC, but mingw-w64 doesn't provide it. It's a pity as we're trying to unify the code for these two targets. I want to retire and surf.

I think I've managed to provide an abstraction over both pthreads and the WinAPI in the `caml_plat_*` namespace. I hope that runtime developers won't have to look elsewhere. We do have to re-implement error-check mutexes to provide the same feature set.

I don't think that @dra27 has had time to review this PR, I would be more comfortable if he approved it explicitly.

gasche commented:
I started a `precheck` job for this PR: https://ci.inria.fr/ocaml/job/precheck/1066/

MisterDA commented:
> I started a `precheck` job for this PR: https://ci.inria.fr/ocaml/job/precheck/1066/

Amazing, it's failing on mingw-w64 both 32 and 64 bits. This is fishy.

```
Makefile.common:492: warning: undefined variable 'CYGWIN'
Makefile.common:494: warning: undefined variable 'MSYS'
```
EDIT: the warning is unrelated to this PR.

```
make[2]: Entering directory '/cygdrive/c/builds/workspace/precheck/flambda/false/label/ocaml-mingw-64/stdlib'
  OCAMLC camlinternalFormatBasics.cmi
make[2]: *** [Makefile:147: camlinternalFormatBasics.cmi] Error 127
make[2]: Leaving directory '/cygdrive/c/builds/workspace/precheck/flambda/false/label/ocaml-mingw-64/stdlib'
make[1]: *** [Makefile:696: coldstart] Error 2
make[1]: Leaving directory '/cygdrive/c/builds/workspace/precheck/flambda/false/label/ocaml-mingw-64'
make: *** [Makefile:855: world.opt] Error 2
```

???


Could you also add the `run-crosscompiler-tests` label?
Somehow the AppVeyor run just stopped too.

MisterDA commented:
I don't understand the multicoretest failure, but since they run on Ubuntu, I don't think they would be affected by the switch from winpthreads to the WinAPI ;)
The AppVeyor failure is due to the build+run of the testsuite exceeding 60 minutes. I see that some of our builds reach 54 minutes, that's unfortunate.
The cross-android build failure is unrelated.

jmid commented:
> I don't understand the multicoretest failure, but since they run on Ubuntu, I don't think they would be affected by the switch from winpthreads to the WinAPI ;) 

I fixed this in https://github.com/ocaml-multicore/multicoretests/pull/567.
You are welcome to just use the `main` branch (compatible with the latest ocaml/ocaml `trunk`) rather than the currently tagged 0.9 https://github.com/ocaml/ocaml/blob/trunk/.github/workflows/multicoretests.yml#L39

MisterDA commented:
Thanks @jmid, multicoretests succeeds on this PR: https://github.com/MisterDA/multicoretests/pull/2!

dra27 commented on .gitmodules:
```diff
@@ -2,7 +2,3 @@
 	path = flexdll
 	url = https://github.com/ocaml/flexdll.git
 	shallow = true
-[submodule "winpthreads"]
```
The winpthreads mentions need to go from `ocaml-variants.opam` as well (see https://github.com/dra27/ocaml/commit/b4d5bc8fe05411e242041b3bac988f7b4a6f9161)

MisterDA replied:
Thanks, I've applied your patch.

dra27 commented on runtime/unix.c:
```diff
@@ -542,3 +542,47 @@ void caml_plat_mem_unmap(void* mem, uintnat size)
+
+/* Threads */
+
+int caml_plat_thread_create(caml_plat_thread *restrict thread,
```
Are these worth doing as `Caml_inline` in the header for Unix instead? The intermediate where this is introduced doesn't build on Windows, which also made me wonder if these functions are worth inlining, just on the basis that it then means that the call sites for the Unix implementation are unchanged?

MisterDA replied:
I agree. I've applied your patch, the `caml_plat_thread_*` family of functions are now `static inline` a.k.a `Caml_inline` in the `<caml/platform.h>` header.

dra27 commented on runtime/caml/platform.h:
```diff
@@ -111,8 +111,11 @@ Caml_inline void cpu_relax(void) {
-#define CAML_PLAT_MUTEX_INITIALIZER SRWLOCK_INIT
+typedef struct {
+  SRWLOCK lock;
+  _Atomic DWORD owner_tid;      /* 0 if not owned */
```
There's a clarifying comment from the 4.x implementation, and I think it's worth having an updated version here:
https://github.com/ocaml/ocaml/blob/8899c05cfc9027e74dfe88c71cc78bac5102e771/otherlibs/systhreads/st_win32.h#L206-L212

MisterDA replied:
I've had a look to winpthreads and ntpl, only winpthreads marks this field as `volatile`, none of them use any sort of lock or atomic access to read or write to it. I'll just add:
> The "owner_tid" field is not always protected by "lock"; it is also accessed without holding "lock".

dra27 commented on runtime/sync_win32.h:
```diff
@@ -45,23 +46,39 @@ Caml_inline int sync_mutex_destroy(sync_mutex m)
 {
-  AcquireSRWLockExclusive(m);
-  return 0;
+  DWORD self_tid = GetCurrentThreadId();
```
The use of `GetCurrentThreadId` resurrects the conversation started in https://github.com/ocaml/ocaml/pull/9846#pullrequestreview-484124107 - there were two concerns. Firstly was whether it was theoretically possible to attack the thread ID and secondly was whether `GetCurrentThreadId` (accessing the TCB and so presumably(?) having to cross into kernel mode) is slower than TLS (which doesn't).

The difference is that we don't have thread descriptors necessarily here - but we could simply have a TLS value in `platform.c` and ensure that the thread creation function sets it? (in this case, the thread creation function obviously mustn't be inlined)

MisterDA replied:
> Firstly was whether it was theoretically possible to attack the thread ID

Are you describing a _Robust Mutex_ from POSIX 2008 (see [`pthread_mutexattr_getrobust(3p)`](https://pubs.opengroup.org/onlinepubs/9799919799/functions/pthread_mutexattr_getrobust.html) and [`pthread_mutexattr_getrobust(3)`](https://man7.org/linux/man-pages/man3/pthread_mutexattr_getrobust.3.html))?
We currently don't use robust mutexes in pthread code.

POSIX states:

> `PTHREAD_MUTEX_STALLED`  
> No special actions are taken if the owner of the mutex is terminated while holding the mutex lock. This can lead to deadlocks if no other thread can unlock the mutex.  
> This is the default value.

The Linux manual says:

> If a mutex is initialized with the `PTHREAD_MUTEX_STALLED` attribute and its owner dies without unlocking it, the mutex remains locked afterwards and any future attempts to call `pthread_mutex_lock(3)` on the mutex will block indefinitely.

I tried to look at [`pthread_mutex_lock`](https://github.com/bminor/glibc/blob/master/nptl/pthread_mutex_lock.c) from NTPL in glibc, and maybe it behaves as the documentation says if the system supports futex, but I don't understand if/how non-robust mutexes are protected in any way from thread id re-use. Errorcheck mutexes probably return `EDEADLK`.

For reference, there's also a thread introducing [Robust mutexes implementation](https://lists.freebsd.org/pipermail/freebsd-threads/2016-May/006018.html) in FreeBSD.

As thread id are global to the system on Windows but unique to each process with POSIX I acknowledge there's a greater risk of re-using a thread id on Windows, but we're also currently not protected against this on POSIX. I suggest we leave this for another adventure. I also note that both in Linux and FreeBSD robust mutex are implemented with support from the kernel. I'm not sure that's something we can reliably do here.

Slightly related, The Old New Thing:
- [What can go wrong if you release an SRWLock from a thread different from the one that acquired it?](https://devblogs.microsoft.com/oldnewthing/20230623-00/?p=108371)
- [Slim reader/writer locks don't remember who the owners are, so you'll have to find them some other way](https://devblogs.microsoft.com/oldnewthing/20110810-00/?p=9933).

dra27 replied:
Slightly hand-wavily, in the 4.x version, the possibly-reused owner was used _after_ the synchronisation primitive (`EnterCriticalSection` then) whereas now it's inspected _before_... I prefer the undefined behaviour we have now, where deadlock still _might_ be erroneously indicated but without having actually messed around with the primitive at all.

In both cases (critical sections and SRW locks), terminating a thread without releasing the lock is UB, so I agree that both then (in #9846) and now I was inadvertently applying a stronger semantics to the Windows mutexes (the actual Windows mutexes I think are robust by that definition, but we've never used them in OCaml because they're far too heavyweight).

dra27 reviewed (commented):
This is really excellent, thank you! The commit series is informative and it would be good to keep that, but the intermediate commits need to build - in particular, that means the winpthread removal needs to happen _after_ the primitives are added, and the commit introducing the thread functions doesn't build on Windows. I had a brief play in https://github.com/dra27/ocaml/commits/winpthreadsectomy which you're welcome to pick anything useful from.

In order to keep the effect minimal on the pthreads version, I suggest inlining the new functions (which I've added in that commit series) - but, given that all the synchronisation primitives have wrapper functions, this is quite possibly not an improvement.

The `GetCurrentThreadId` part in the errorcheck implementation I think is worth looking at, because I hope the fix is easy - in the Windows version, platform.c needs to reserve a TLS slot at startup and then the Windows version of `caml_plat_thread_create` needs to ensure that that TLS variable is set to the thread ID (or to an atomically incremented private counter of our own) as soon as the thread is created.

dra27 commented:
As an aside, what do we need to do separately to get the multicoretests to pass in this PR?

jmid commented:
> As an aside, what do we need to do separately to get the multicoretests to pass in this PR?

I fixed this in [ocaml-multicore/multicoretests#567](https://github.com/ocaml-multicore/multicoretests/pull/567). To avoid changing the test suite under your feet, we originally went with running a tagged version (presently `0.9`). You are welcome to just use the `main` branch (compatible with the latest `trunk`) instead https://github.com/ocaml/ocaml/blob/trunk/.github/workflows/multicoretests.yml#L39.

I'll roll a `0.10` next week and send a PR with an update.

dra27 commented:
Ah, thanks @jmid (I was being slow joining the various dots together) - for now, I've pushed the sha of the merge of that PR as the ref, so no pressing rush for a 0.10 release.

jmid commented:
Note that the ocaml/ocaml GHA workflow only tests under Linux with/without the debug runtime.
To torture the concurrency primitives under Windows @MisterDA has given the PR a spin here https://github.com/MisterDA/multicoretests/tree/winpthreadsectomy
where it holds up nicely too :slightly_smiling_face:

MisterDA commented:
> The `GetCurrentThreadId` part in the errorcheck implementation I think is worth looking at, because I hope the fix is easy - in the Windows version, `platform.c` needs to reserve a TLS slot at startup and then the Windows version of `caml_plat_thread_create` needs to ensure that that TLS variable is set to the thread ID (or to an atomically incremented private counter of our own) as soon as the thread is created.

> secondly was whether `GetCurrentThreadId` (accessing the [Thread Control Block](https://en.wikipedia.org/wiki/Thread_control_block) and so presumably(?) having to cross into kernel mode) is slower than TLS (which doesn't).

> The difference is that we don't have thread descriptors necessarily here - but we could simply have a TLS value in `platform.c` and ensure that the thread creation function sets it? (in this case, the thread creation function obviously mustn't be inlined)

According to:
- https://stackoverflow.com/questions/15007771/is-operation-of-getting-id-of-current-thread-time-expensive
- https://stackoverflow.com/questions/12909215/is-getting-the-thread-id-expensive-in-terms-of-performance

`GetCurrentThreadId` could? should? be fast ([two assembly instructions!](https://github.com/lhmouse/mcfgthread/blob/0c899f075d6674f0f6e1b8bec8d8532a26861a5c/mcfgthread/thread.h#L349-L357)) because it only accesses a known offset inside the [Thread Information Block](https://en.wikipedia.org/wiki/Win32_Thread_Information_Block).

> It is accessed from the FS [segment register](https://en.wikipedia.org/wiki/X86_memory_segmentation) on 32-bit Windows and GS on 64-bit Windows.

 It is also similarly used in [winpthreads](https://github.com/mingw-w64/mingw-w64/blob/a0fd21158e9f3d625dd6d682443bf9c905ff5e01/mingw-w64-libraries/winpthreads/src/mutex.c#L120-L146) and [Firefox](https://github.com/mozilla-firefox/firefox/blob/main/mozglue/misc/WinUtils.h); so it seems we can keep it. We could also provide an `inline` version, if you're _that_ worried about performance ;)

dra27 commented:
> It is also similarly used in [winpthreads](https://github.com/mingw-w64/mingw-w64/blob/a0fd21158e9f3d625dd6d682443bf9c905ff5e01/mingw-w64-libraries/winpthreads/src/mutex.c#L120-L146) and [Firefox](https://github.com/mozilla-firefox/firefox/blob/main/mozglue/misc/WinUtils.h); so it seems we can keep it. We could also provide an `inline` version, if you're _that_ worried about performance ;)

Agreed - the performance was secondary to the (I think now, incorrect) worry about re-use!

nojb commented:
What is the status of this? Given that there's consensus for it in principle, I wouldn't want it to be forgotten :)

MisterDA committed:
Introduce a caml_plat abstraction for threads

Introduce new caml_plat_thread and caml_plat_thread_attr to abstract
over the system's thread id/handle type and thread attributes.

MisterDA committed:
Duplicate sync_posix.h into sync_win32.h

Done to ease the diff when switching to WinAPI primitives.

MisterDA committed:
Regroup definitions using pthreads in platform.h

MisterDA committed:
Duplicate functions using pthreads into unix and win32 backends

Done to ease the diff when switching to WinAPI primitives.

MisterDA committed:
Inline the content of st_pthreads.h into st_win32.h

Substitute then remove the copyright header.

    sed -i \
      -e '/#include "st_pthreads.h"/{r otherlibs/systhreads/st_pthreads.h' \
      -e 'd}' \
      otherlibs/systhreads/st_win32.h

Done to ease the diff when switching to WinAPI primitives.

MisterDA committed:
Duplicate definitions using pthreads in platform.h

Done to ease the diff when switching to WinAPI primitives.

MisterDA committed:
Implement concurrency primitives using WinAPI

Don't include pthread.h on Windows to avoid using pthreads emulation
libraries (such as winpthreads) by mistake.

Functions from platform.c that used pthreads are moved to unix.c,
their Windows counterparts are implemented in win32.c. Introduce a new
sync_win32.h header, counterpart of sync_unix.h. Use the
caml_plat_thread and caml_plat_thread_attr to abstract over the
system's thread id/handle type and thread attributes.

Use pairs of SRWLOCK and CONDITION_VARIABLES to replace pthreads
mutexes and condition variables.

POSIX states:

> The pthread_mutex_lock() function shall fail if:
> [EDEADLK]
>   The mutex type is PTHREAD_MUTEX_ERRORCHECK and the current thread
>   already owns the mutex.

The Windows documentation states:

> Exclusive mode SRW locks cannot be acquired recursively. If a thread
> tries to acquire a lock that it already holds, that attempt will
> [...] deadlock (for AcquireSRWLockExclusive).

Deadlocks caused by recursive attemps to lock cannot be detected on
Windows.

MisterDA committed:
Disable winpthreads compilation and remove the winpthreads submodule

MisterDA committed:
Implement errorcheck mutexes on Windows

MisterDA commented:
I have rebased the PR on the current trunk, using @dra27 patches and suggestions. I've also added new commits that just duplicate the pthread code into the Windows-specific parts so that the diff when switching from pthreads to the Windows API is easier to understand.
~Not sure why hygiene rejects the copyright header.~

nojb commented on runtime/sync_win32.h:
```diff
@@ -0,0 +1,133 @@
+/*                                 OCaml                                  */
+/*                                                                        */
+/*          Xavier Leroy and Damien Doligez, INRIA Rocquencourt           */
+/*   Copyright 1996 Institut National de Recherche en Informatique et     */
```
Missing empty line between attribute and copyright?

MisterDA replied:
Thanks!

dra27 commented:
@MisterDA is just checking fixes to the tests, but just a note from me for this not to merged just yet - AppVeyor is giving a false positive for the build which I’m just investigating and fixing

dra27 committed:
Check in CI that libwinpthread-1.dll is not linked

dra27 committed:
Sanitise the AppVeyor environment

In particular, remove utilities which installed mingw-w64 runtime DLLs
that wouldn't otherwise be in PATH.

dra27 commented:
The "problem" on the mingw-w64 port is that because we're no longer using pthreads, `-l:libpthread.a` is ignored by the linker because it's not required. However, GCC synthesises calls to pthreads functions in order to implement TLS and that then causes the _default_ `-lpthread` to link the DLL, which is the failure I've now surfaced in AppVeyor. We weren't seeing it before because there's other software in the CI image which puts the winpthreads runtime DLL into the PATH - I've added a check both to ensure that nothing built links with `libwinpthread-1.dll` and an additional hardening to remove entries from `PATH` which add it. As it stands, this excellent PR therefore accidentally regresses #11586.

There are various solutions to this - sledgehammers include using `-static` or wrapping the `-l:libpthread.a` with a `-Wl,--whole-archive`...`-Wl,--no-whole-archive` "bracket" (`-Wl,-Bstatic`...`-Wl,-Bdynamic` can't work here). My friend @claude pointed out the `--undefined` linker option (using `-Wl,--undefined=pthread_getspecific -l:libpthread.a`) and fact-checking that led me to `--require-defined`. The trick there is to make the linker pretend it needs `pthread_getspecific` when it considers `-l:libpthread.a` and so that causes `thread.o` to be linked (as before) and then GCC's pthreads function calls are already linked when we reach the implicit `-lpthread`. There are two problems with that:
1. The option has to get through `flexlink`, which is fiddly
2. It's only needed when linking executables which use the runtime (i.e. `ocamlrun`, etc.) and _not_ `ocamlyacc`, `sak`, etc. At present `OC_LDFLAGS` is an "all or nothing", and this approach would require catering for `OC_LDFLAGS`-for-the-runtime; `OC_LDFLAGS`-for-`MKEXE_VIA_CC`-for-ocamlruns and `OC_LDFLAGS`-for-`MKEXE_VIA_CC`-for-everything-else all of which makes my head hurt

So I propose a simpler, if slightly dirty, approach in `startup_aux.c` - on mingw-w64 only, we generate a call to the function `pthread_getconcurrency` in `caml_startup_aux` (this just returns an integer which we ignore). `caml_startup_aux` is used by both startup\_nat.c and startup\_byt.c, so both runtimes will then contain a _single_ reference to a pthreads function on mingw-w64. I debated putting it in `caml_init_domains`, but I think startup\_aux.c is better precisely because it _doesn't_ contain any threading stuff - i.e. it's not likely ever to have any other pthreads-related stuff added by accident (where domains.c obviously could).

From a cursory reading, I think mingw-w64's GCC has two models for threading - one using pthreads (guaranteed POSIX compliance) and one using native Win32 (therefore with caveats). There are two reasons I think we should go with letting GCC use pthreads here:
1. It's the default
2. Regardless of what `ocamlrun` does, _user C code_ may also end up synthesising this these calls, and I think it creates both a better experience in general, but also it's more compatible with what 5.0-5.4 are implicitly all doing

dra27 committed:
Ensure mingw-w64 still links pthreads statically

mingw-w64 GCC synthesises some calls to pthreads functions to implement
TLS. Ensure that the runtime still links pthreads statically on
mingw-w64 by including a single call to a pthreads function in the
runtime itself which causes the required module of libwinpthreads to be
linked for GCC, and thus avoids executables suddenly requiring the
libwinpthreads runtime DLL.

dra27 reviewed (approved):
LGTM, thank you @MisterDA! It's running through [precheck#1093](https://ci.inria.fr/ocaml/job/precheck/1093/) hopefully one last time. I'm assuming that the Windows multicoretests aren't throwing up anything (I removed the label here as the known failure in #14349 is adding noise).

This is ready to go, as long as both @MisterDA and @nojb are happy with 7ca57cf. That commit could be moved before 864e789 ("Implement concurrency primitives using WinAPI") so that all stages of the commit series compile (I wouldn't squash that into the removal commit itself).

MisterDA commented:
Thanks for this amazing detective work, it would have probably taken me another year ;) I think your explanations and commits make sense.

> That commit could be moved before https://github.com/ocaml/ocaml/commit/864e7893a83ea54f41ec8b14cac5f3d6b12d4b2b ("Implement concurrency primitives using WinAPI") so that all stages of the commit series compile (I wouldn't squash that into the removal commit itself).

I've moved the commit.

dra27 commented:
There's one last thing which precheck has thrown up - the i686 mingw-w64 build is broken, but that's actually because the test for `SetThreadDescription` in #13504 is broken. I'll open a separate PR for that in a second - but I'm wondering if you want to ensure that the pthreads functions are never used for Windows in `caml_set_current_thread_name`?

nojb commented:
@MisterDA @dra27: CI is fully green. #14423 notwithstanding, is the PR good to go? If yes, I suggest we push the button.

MisterDA committed:
Ensure pthreads functions are never used for Windows

in caml_set_current_thread_name.

MisterDA commented:
> ensure that the pthreads functions are never used for Windows in `caml_set_current_thread_name`

Good idea, done!

> is the PR good to go?

Maintenant, oui.

> If yes, I suggest we push the button.

🚀

dra27 committed:
Accelerate AppVeyor

Use a similar trick as on GitHub Actions - build with `-j` but if the
build fails, repeat without -j and then exit with error.

dra27 commented:
Slightly hot-fixing AppVeyor-in-a-PR...

gasche commented:
I have a naive question, which should of course not be considered blocking for this PR: if I correctly understand from @dra27's scary story that mingw-w64 uses gcc and carries a full `pthreads` implementation, why not use `pthreads` in that configuration? Are there clear benefits to using our own reimplementation on top of the Win32 API when pthreads is available? (Maybe the pthreads is an emulated layer that has worse performance, for example?)

gasche commented:
(Github says that "this branch has conflicts that must be resolved".)

dra27 commented:
> (Github says that "this branch has conflicts that must be resolved".)

Alas, the opam file conflicted with changes #14246

dra27 commented:
It's a good question, @gasche, and it doesn't have a 100% clear answer:
- The winpthreads implementation we were using before _is_ the mingw-w64 pthreads implementation, indeed - we ("we" here is mainly @MisterDA!) were maintaining the build of that with MSVC (with upstream agreement, thankfully)
- The core premise of this change (as @stedolan has also said in the past) is that OCaml's requirements are semantically much lighter-weight than pthreads requires - in practice, that means that the implementation @MisterDA has done here is simpler than the winpthreads implementations of the primitives we were using before. That's partly down to not needing to implement things we don't need and also down to not needing to support quite so many versions of Windows, enabling us to take advantage of some relatively modern native concurrency primitives
- I _think_ it's better to have consistency in our implementation between the MSVC and mingw-w64 ports, given all that. In this case, we're confident that the implementations here are semantically what we need for all the other concurrency primitives. What we're then allowing is a much smaller part of it as part of pthreads to come in just for GCC's implementation of TLS. The code in threads.c for all of this is unsurprisingly close to what we do anyway (TLS requires OS support, after all), and GCC here is implementing part of C11, i.e. it's implementing a highly portable C standard _using_ winpthreads rather than the other parts of winpthreads which are _emulating_ the much less portable POSIX standard pthreads.h
- If any of those assumptions prove wrong in future, we could choose to change the mingw-w64 port either not to use pthreads at all (I think) or to use winpthreads completely, as it does before this PR - i.e. this PR doesn't close the door on any of the options for mingw-w64, the main thing is that it simplifies MSVC maintenance _a lot_ (and, anticipating a follow-up question, clang doesn't help us here - they'll have implemented TLS without using winpthreads, as that's part of mingw-w64 GCC!)

I hope that's a bit less unclear - it's kinda of a brain-dump of my internal monologue on it 😊

dra27 commented:
Thanks, @MisterDA!
